spicy/toolchain/src/compiler/codegen/parser-builder.cc

```diff
@@ -109,6 +109,34 @@ struct ProductionVisitor
     hilti::util::Cache<std::string, ID> parse_functions;
     std::vector<hilti::declaration::Field> new_fields;
     std::vector<Expression> _destinations;
+    std::vector<ID> _path; // paths of IDs followed to get to current unit/field
+
+    // RAII helper to update the visitor's `_path` as we descend the parse tree.
+    class PathTracker {
+    public:
+        PathTracker(ProductionVisitor* v, const ID& id) : v(v) { v->_path.emplace_back(id); }
+        PathTracker() = delete;
+        ~PathTracker() {
+            if ( v )
+                v->_path.pop_back();
+        }
+
+        PathTracker(const PathTracker& other) = delete;
+        PathTracker(PathTracker&& other) noexcept {
+            v = other.v;
+            other.v = nullptr;
+        }
+
+        PathTracker& operator=(const PathTracker& other) = delete;
+        PathTracker& operator=(PathTracker&& other) noexcept {
+            v = other.v;
```

Comment:
Can we make at least this one safe against self-assignment, e.g.,

```suggestion
            if ( this == &other )
                return *this;

            v = other.v;
```

I think for the move ctor above the only we could do would e.g., `assert(this != &other);`.